### PR TITLE
[air] Add streaming BatchPredictor support

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Add pipelined batch prediction support. This technically isn't required, but I think it is a bit discoverability improvement to surface it as a top-level API instead of relying on the user to know how to generate pipelines.

### CODE REVIEW

**Comment by ericl** [comment]
@richardliaw @maxpumperla any idea why the doctest is failing with a MagicMock error? It runs fine locally with `python -m doctest file`

```


  | File "../../python/ray/air/batch_predictor.py", line ?, in default
  | Failed example:
  | batch_pred = BatchPredictor(
  | Checkpoint.from_dict({"x": 0}), DummyPredictor)
  | Exception raised:
  | Traceback (most recent call last):
  | File "/opt/miniconda/lib/python3.6/doctest.py", line 1330, in __run
  | compileflags, 1), test.globs)
  | File "<doctest default[4]>", line 2, in <module>
  | Checkpoint.from_dict({"x": 0}), DummyPredictor)
  | File "/ray/python/ray/air/batch_predictor.py", line 33, in __init__
  | self.checkpoint_ref = checkpoint.to_object_ref()
  | File "/ray/python/ray/air/checkpoint.py", line 287, in to_object_ref
  | return ray.put(self.to_dict())
  | File "/ray/python/ray/_private/client_mode_hook.py", line 100, in wrapper
  | if client_mode_should_convert(auto_init=auto_init):
  | File "/ray/python/ray/_private/client_mode_hook.py", line 124, in client_mode_should_convert
  | ray.init()
  | File "/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
  | return func(*args, **kwargs)
  | File "/ray/python/ray/worker.py", line 1257, in init
  | head=True, shutdown_at_exit=False, spawn_reaper=True, ray_params=ray_params
  | File "/ray/python/ray/node.py", line 273, in __init__
  | self.start_head_processes()
  | File "/ray/python/ray/node.py", line 1095, in start_head_processes
  | self._write_cluster_info_to_kv()
  | File "/ray/python/ray/node.py", line 1073, in _write_cluster_info_to_kv
  | ray_usage_lib.put_cluster_metadata(self.get_gcs_client(), NUM_REDIS_GET_RETRIES)
  | File "/ray/python/ray/_private/usage/usage_lib.py", line 382, in put_cluster_metadata
  | num_retries=num_retries,
  | File "/ray/python/ray/_private/utils.py", line 1283, in internal_kv_put_with_retry
  | key, value, overwrite=True, namespace=namespace
  | File "/ray/python/ray/_private/gcs_utils.py", line 138, in wrapper
  | return f(self, *args, **kwargs)
  | File "/ray/python/ray/_private/gcs_utils.py", line 249, in internal_kv_put
  | f"Failed to put value {value!r} to key {key!r} "
  | RuntimeError: Failed to put value b'{"ray_version": "3.0.0.dev0", "python_version": "3.6.13"}' to key b'CLUSTER_METADATA' due to error <MagicMock name='mock.gcs_service_pb2_grpc.InternalKVGcsServiceStub().InternalKVPut().status.message' id='140197865852264'>
 ```

**Comment by ericl** [comment]
Will skip the doctests for now.

File: python/ray/air/batch_predictor.py [hunk 1]
```diff
@@ -102,3 +122,69 @@ def __call__(self, batch):
...
         )
+
+    def predict_pipelined(
+        self,
+        data: ray.data.Dataset,
+        *,
+        blocks_per_window: Optional[int] = None,
+        bytes_per_window: Optional[int] = None,
+        **kwargs,
+    ) -> ray.data.Dataset:
```
Comment (richardliaw) [inline python/ray/air/batch_predictor.py:133]: this signature return should be a Pipeline right?

**Review by krfricke — APPROVED** [review]
Nice!

**Review by richardliaw — APPROVED** [review]

**Pull request merged**